water/lib/Transforms/MemrefDecomposition.cpp

```diff
@@ -0,0 +1,653 @@
+static std::tuple<LogicalResult, Value, SmallVector<OpFoldResult>,
+                  SmallVector<OpFoldResult>>
+unflattenDescriptor(ValueRange values, MemRefType memrefType) {
+  unsigned rank = memrefType.getRank();
+  if (values.size() != 1 + rank * 2)
+    return {failure(), {}, {}, {}};
+
+  int64_t staticOffset = 0;
+  SmallVector<int64_t> staticStrides;
+  if (failed(memrefType.getStridesAndOffset(staticStrides, staticOffset)))
+    return {failure(), {}, {}, {}};
+
+  Value buffer = values.front();
+  values = values.drop_front();
+  ValueRange sizes = values.take_front(rank);
+  values = values.drop_front(rank);
+  ValueRange strides = values.take_front(rank);
+
+  SmallVector<OpFoldResult> mixedSizes;
+  SmallVector<OpFoldResult> mixedStrides;
+  auto intType = IndexType::get(memrefType.getContext());
+  for (auto i : llvm::seq(rank)) {
+    int64_t size = memrefType.getDimSize(i);
+    if (ShapedType::isDynamic(size))
+      mixedSizes.push_back(sizes[i]);
+    else
+      mixedSizes.push_back(IntegerAttr::get(intType, size));
+
+    if (ShapedType::isDynamic(staticStrides[i]))
+      mixedStrides.push_back(strides[i]);
+    else
+      mixedStrides.push_back(IntegerAttr::get(intType, staticStrides[i]));
+  }
+
+  return {success(), buffer, mixedSizes, mixedStrides};
+}
+
+static MemRefType make0DMemRefType(MemRefType type) {
+  return MemRefType::get({}, type.getElementType(), MemRefLayoutAttrInterface{},
+                         type.getMemorySpace());
+}
+
+static Type getMemrefStructType(OpBuilder &builder, Location loc, Type ptrType,
+                                unsigned rank) {
+  auto i64 = builder.getIntegerType(64);
+  if (rank == 0)
+    return LLVM::LLVMStructType::getLiteral(builder.getContext(),
+                                            {ptrType, ptrType, i64});
+
+  auto arrayType = LLVM::LLVMArrayType::get(i64, rank);
+
+  return LLVM::LLVMStructType::getLiteral(
+      builder.getContext(), {ptrType, ptrType, i64, arrayType, arrayType});
+}
+
+/// Extract the pointer from a memref descriptor.
+static Value toPtr(OpBuilder &builder, Location loc,
+                   LLVM::LLVMPointerType ptrType, Value value) {
+  auto memrefType = cast<MemRefType>(value.getType());
+  auto memrefStructType =
+      getMemrefStructType(builder, loc, ptrType, memrefType.getRank());
+  value =
+      UnrealizedConversionCastOp::create(builder, loc, memrefStructType, value)
+          .getResult(0);
+  return MemRefDescriptor(value).alignedPtr(builder, loc);
+}
+
+/// Create a 0D memref descriptor from a pointer.
+static Value fromPtr(OpBuilder &builder, Location loc, MemRefType memrefType,
+                     Value value) {
+  auto ptrType = cast<LLVM::LLVMPointerType>(value.getType());
+  assert(memrefType.getRank() == 0 && "only 0D memrefs supported");
+
+  auto memrefStructType = getMemrefStructType(builder, loc, ptrType, 0);
+  auto descriptor = MemRefDescriptor::poison(builder, loc, memrefStructType);
```

Comment:
Nit: please expand `auto` unless the type is obvious from statement-level context.